### PR TITLE
TYP: make the type annotations of read_csv & read_table discoverable

### LINKED ISSUE #25648: ERR: read_csv exposes an internal function when bad argument is specified
When a user tries to specify a non-existent argument in `read_csv` the error message pertains to what seems to be an internal function (`parser_f`).  This behavior might be a bit surprising / confusing to users, who would expect the error message to refer to the function actually being used.

```python
import pandas as pd
df = pd.read_csv("some_file.csv", bad_arg=True)
```

yields

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-d4606a60902f> in <module>()
----> 1 df = pd.read_csv("some_file.csv", bad_arg=True)

TypeError: parser_f() got an unexpected keyword argument 'bad_arg'
```

### PR DESCRIPTION
**Pull request by topper-123**
- [x] closes #25648
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

In master the type of ``read_csv`` and ``read_table`` is 'Any'.  This PR makes the functions signatures discoverable by mypy. Also fixes the error message when wrong kwarg is passed.

### CODE REVIEW

**Comment by WillAyd** [comment]
I am on board with this. This probably also closes #25648

**Comment by WillAyd** [comment]
So maybe worth pulling the test case from #33023

**Comment by WillAyd** [comment]
@gfyoung

**Comment by gfyoung** [comment]
I'm a bit torn on this TBH

I see the benefits to this change, but on the other hand, it does mean we will need to maintain two copies of essentially the same signature.  This function generator setup was meant to alleviate that.

If we're mostly okay with maintaining two copies of the signature, then go for it.  @pandas-dev/pandas-core

**Comment by topper-123** [comment]
@WillAyd , good idea taking in that test from #33023. I'll do that in the next commit.

@gfyoung , that was the reason I added the test ``test_read_table_same_signature_as_read_csv``, to ensure the signatures don't drift apart :-)

**Comment by gfyoung** [comment]
@topper-123 : Ah, that's a fair point.  Okay...it still feels little hacky, but the test does make me feel about this.

**Review by gfyoung — APPROVED** [review]

File: pandas/io/parsers.py [hunk 1]
```diff
@@ -530,176 +530,229 @@ def _read(filepath_or_buffer: FilePathOrBuffer, kwds):
...
-    parser_f.__name__ = name
-
-    return parser_f
+    if engine is not None:
+        engine_specified = True
+    else:
+        engine = "c"
+        engine_specified = False
+
+    kwds.update(
```
Comment (simonjayhawkins) [inline pandas/io/parsers.py:635]: at some point, would be beneficial for consistency checking to not use the dictionary here as mypy doesn't check this.

**Review by simonjayhawkins — APPROVED** [review]
Thanks @topper-123 lgtm.

**Review by jreback — APPROVED** [review]
any comments @TomAugspurger or @jorisvandenbossche

**Comment by TomAugspurger** [comment]
No comments.

**Comment by WillAyd** [comment]
Thanks @topper-123

**Pull request merged**